### ISSUE TITLE
Redesign ALPN mismatch handling

### ISSUE THREAD

**Issue by bemasc** [MikeBishop/dns-alt-svc#73]
Currently, if a server publishes a QUIC-only HTTPSSVC RRSet with ESNI, there is no way for a client to fall back to a non-QUIC connection, because doing so would reveal the SNI.  This could increase the likelihood of partial outages for server admins who haven't considered the small fraction of users whose network path does not support QUIC.

Reviewers have reported concerns that this creates an undesirable level of fragility.  We should consider whether there is an alternative design that would be less likely to result in accidental breakage.

**Comment by bemasc** [comment]
After some thought, I'm leaning towards just declaring that if an "h3" attempt fails you should (may?) fall back to "h2" with the rest of the parameters unchanged.  That's a hack, and it won't always work (e.g. when H3 and H2 aren't colocated), but it avoids the "footgun" where an ordinary H3+H2 server becomes QUIC-only due to a server admin publishing an H3-only HTTPSSVC RRSet.

**Comment by bemasc** [comment]
I've just learned of another closely related case: HTTP/1.1-only clients.  Just like an HTTP/2 client can't use an RR that has `alpn=h3`, an HTTP/1.1-only client can't use an RR that has `alpn=h2`.  This seems like a common mistake that people might make, but there are significant use cases for HTTP/1.1 with HTTPSSVC (e.g. WebSocket, which is currently mostly HTTP/1.1-only despite RFC 8441).

As with QUIC fallback, this can be handled within the current draft, but it requires any server that uses ESNI to remember to publish RRs for both `alpn=http/1.1` and `alpn=h2` (and `alpn=h3` if they do QUIC) in each RRSet.  Failure to include an `alpn=http/1.1` RR will cause silent loss of ESNI protection for HTTP/1.1 clients, which seems like a footgun.

We should think a little bit harder about what the ALPN field means, and whether there's a way to convey the right information (mostly whether this endpoint supports QUIC) in a way that makes misconfiguration harder.

**Comment by bemasc** [comment]
New plan: `proto` (I promise I support) and `no_proto`* (I don't support so don't bother trying).  For HTTPSSVC, "http/1.1" and "h2" are in the default set.  Multiple keys are allowed.

Both share the ALPN registry.

*Need to add "_" to the allowed characters.

**Comment by davidben** [comment]
How does this interact with someone adding a new ALPN token that's used out of the TCP + TLS + {http/1.1, h2} entrypoint? What is the mapping to Alt-Svc, given that Alt-Svc got ALPN wrong? How does one handle the HTTPSSVC http/1.1 vs h2 subsetting getting out of sync with the server?

The root problem here is we're trying to repurpose ALPN, an in-band mid-connection-establishment variant selector, for out-of-band pre-connection-establishment entrypoint selector. These two objects don't have the same type. Specifying either "http/1.1" or "h2" in there doesn't make sense. Rather, we should add a name for the TCP + TLS + {any HTTPS variant} stack to stick in there and drop the in-band vs. out-of-band interactions.

**Comment by enygren** [comment]
One option may be to share the ALPN registry but be explicit about the differences here.  (ie, require each of the tokens to have a semantic meaning defined here.)

Should clients only offer (or only accept server offers) of ALPN tokens in the set "allowed" by the HTTPSSVC set?  Or is that counter-productive?

**Comment by davidben** [comment]
> Should clients only offer (or only accept server offers) of ALPN tokens in the set "allowed" by the HTTPSSVC set? Or is that counter-productive?

Right, this mess is why Alt-Svc's use of ALPN was wrong. If we say yes here, we break downgrade protection of the protocol negotiation. That's a nonstarter. If we say no here, we have to worry about what happens if the server picks something outside of the set. Alt-Svc says it's an error, which means instead we have a deployment hazard. The PR just says anything goes. Ambiguity breeds interoperability problems, so that's no good. The spec should say what it means. We can't say anything in the same entrypoint is good because that's not defined. It's also bizarre because that means "http/1.1" and "h2" are just aliases for the thing we've failed to define.

**Comment by enygren** [comment]
That might be a good argument for disconnecting these protocols from ALPN and having a separate registry for them with definitions that are independent of Alt-Svc?  I think what we want is:
* Extensible (ie, can add new protocols down the road)
* Able to have service endpoint support a subset of protocols.  (eg, HTTP/3 using unstable anycast IPs because the connection IDs are big enough to fix up and reforward during route changes, whereas TCP-based H/1.1 and H2 stick somewhere more stable).  Or another use-case is rolling out a new protocol to a subset of endpoints.
* Minimize deployment hazards of all forms.
* Have good downgrade protection (especially given DNS is often unauthenticated)
* ... (maybe some other properties?)

**Comment by davidben** [comment]
(I meant to do a proper writeup of this issue this week but didn't end up having time. :-( Ah well, discussion on GitHub works.)

I think separate registry or just overloading the existing one works. Separate is probably cleaner, but allocating new tokens may be easier to wedge back into Alt-Svc. I do want to touch on two of your points though, deployment hazards and downgrade protection. I think these attributes may be somewhat forced by whether the negotiation is in-band or out-of-band.

ALPN, in its original form, is an in-band protocol variant selector. It's negotiated in the TLS handshake, which gives us nice properties:
* Deployment concerns are minimal. A single server can unilaterally reconfigure itself without consequences, so gradual rollout, rollback, etc., is natural.
* Downgrade protection is natural. TLS promises that the protocol you get is the one you should have gotten.

This makes this a natural pattern for HTTP/2.

QUIC can't use this trick. It needs an out-of-band signal, perhaps in an HTTP header (`Alt-Svc`, formerly `Alternate-Protocol`) or the DNS (this draft). Both of these involve some client state (caching, etc.). This has analogous consequences:
* The server's configuration may get out of sync with the client's state. A priori, this means connection failure, so we must handle this. We pair QUIC with a TCP fallback (or run in parallel), which we need anyway due to network intolerance. Corollary: the TCP fallback is critical.
* We do not have natural downgrade protection. An attacker can cause us to negotiate the "wrong" one of the two. Corollary: the QUIC and TCP flows must be capable of comparable security.

I may be overgeneralizing from these examples, but these two properties seem closely related. Downgrade protection means you selected the thing you were supposed to, but if your selection is based on possibly stale information, it's hard to have downgrade protection. (Unless maybe the server reacts to your stale information and has an in-band fixup operation, in which case the out-of-band signal is more a hint than a selection mechanism.)

Anyway, I think this framing helps illustrate how different of animals TLS/ALPN and HTTPSSVC/ALPN are.

**Comment by bemasc** [comment]
I'm OK with replacing "h2" with "tls" in HTTPSSVC, but I want to point out that this would cause the spec to lean _more_ heavily on clients having a full implementation of fallback.

Consider a future CDN that drops HTTP/1.1 support for some of its fleet.  In the scheme of PR #89, endpoints without HTTP/1.1 support are clearly marked, so clients that need HTTP/1.1 can identify the relevant endpoints before the connection attempt.  If instead of (h2, h3) the protocol set was (tls, quic), clients would have to attempt the connection before learning that "http/1.1" is not supported.

To generalize, whenever a client might be compatible with only a subset of endpoints, the client must either be able to tell this from the DNS record or must implement fallback.  The list of ALPNs isn't necessary for enabling the client to attempt a connection to an endpoint, but it is useful to allow clients to avoid attempts that will fail.

**Comment by davidben** [comment]
I don't think that holds. Rather, I think we stop pretending to support a deployment doing that. If you're ready to drop HTTP/1.1 support (maybe you only care about modern clients), go ahead and drop it. If you're not, your fleet should still support HTTP/1.1.

You could make the same set of statements about a server deployment that wants to disable TLS 1.2 on some machines but not others. Or a server deployment that wants a completely different set of TLS cipher suite capabilities on each machine. With unbounded complexity we could express and handle support unbounded partial deployments, but that doesn't mean it's a good idea.

**Comment by MikeBishop** [comment]
In the eventual world where QUIC has version negotiation and ALPN happens in-band for both TCP and UDP, most of what you really need is an indication of TCP vs. UDP and let TLS take care of the specifics at connection-time.  However, a hint about what is supported can make things faster, especially in the QUIC case.

The only case that doesn't cover is when you don't speak any of the protocol variants supported by that endpoint.  The current system lets you discover that without trying a connection.

**Comment by davidben** [comment]
An ALPN hint is somewhere between not useful at all and barely useful. I say barely useful because the one-bit "only HTTP/1.1" vs "supports something not HTTP/1.1" does allow us to avoid opening some connections that we'd otherwise throw away if, say, 10 HTTP requests enter the net stack at once.

However, if you want an ALPN or QUIC version hint, the needs to be explicit that it is a hint. The key thing about hints is that they are non-binding. They do not impact how an individual connection is negotiated, which is important for both the downgrade protection and deployability aspects above.

In contrast, the TCP+TLS vs QUIC indication is not a hint. It fundamentally impacts how an individual connection is negotiated. It does not have the downgrade protection and deployability properties. And if you want to support separate IPs for HTTP/2 and HTTP/3, it cannot be a hint.

These two use cases cannot be mixed up. Otherwise we land in exactly this problem. Any solution here *must* split the two apart.

**Comment by davidben** [comment]
Actually, here's a stronger reason why having different ALPN tokens on different parts of your fleet is no good. Suppose the CDN wanted to have some h1-only machines and some h2-only machines. Even if I get directed to the right one via HTTPSSVC, the network can redirect my packets to the h1-only machines and I get force down to h1.

The security model for multi-homed TLS services has always that we assume all instances of a service are equivalent and downgrade protection is limited to the weakest instance. Due to deployment requirements, instances may have different configurations temporarily, but we require it eventually converges over time.

Thus these kinds of uneven deployment games are only acceptable in scenarios where we don't care about downgrade protection (TCP vs QUIC) because we met the same needs in other ways (QUIC must be able to meet TCP's security level).

**Comment by bemasc** [comment]
I've opened a new PR to explore a version where we negotiate the transport up front, instead of the ALPN: https://github.com/MikeBishop/dns-alt-svc/pull/97
